File: src/main/scala/work/chiro/game/prop/BulletProp.scala
package work.chiro.game.prop

import work.chiro.game.GlobalConfigLoader.config
import work.chiro.game.aircraft.{AbstractAircraft, HeroAircraft}
import work.chiro.game.animate.{AnimateContainer, AnimateLinearToTarget, AnimateVectorType}
import work.chiro.game.application.{ImageResourceFactory, Main}
import work.chiro.game.basic.PositionType.Position
import work.chiro.game.basic.Vec2Double
import work.chiro.game.logger
import work.chiro.game.utils.getTimeMills

class BulletProp(posInit: Position, animateContainer: AnimateContainer[Vec2Double])
  extends AbstractProp(posInit, animateContainer) {
  val bloodIncrease = 100

  override def handleAircrafts(enemyAircrafts: List[AbstractAircraft]) = {
    logger.info("FireSupply active!")
    HeroAircraft.getInstance.increasePower()
  }

  override def create(position: Position) = BulletProp.create(position)

  override def getImage = BulletProp.getImage
}

object BulletProp extends ImageResourceFactory {
  override def getImageCachedPath = "images/prop_bullet.png"

  def create(position: Position) = {
    new BulletProp(
      position, new AnimateContainer[Position](List(
        new AnimateLinearToTarget(position, new Position(position.getX, config.window.playHeight),
          AnimateVectorType.PositionLike.id, getTimeMills, 20000)
      ))
    )
  }
}

File: src/main/java/work/chiro/game/prop/BulletProp.java
package work.chiro.game.prop;

import work.chiro.game.aircraft.AbstractAircraft;
import work.chiro.game.aircraft.HeroAircraftFactory;
import work.chiro.game.animate.AnimateContainer;
import work.chiro.game.application.Game;
import work.chiro.game.vector.Vec2;

import java.util.List;

/**
 * @author Chiro
 */
public class BulletProp extends AbstractProp {
    public BulletProp(Vec2 posInit, AnimateContainer animateContainer) {
        super(posInit, animateContainer);
    }

    @Override
    public void handleAircrafts(List<AbstractAircraft> enemyAircrafts) {
        playSupplyMusic();
        System.out.println("FireSupply active!");
        HeroAircraftFactory.getInstance().increaseShootNum();
        Game.getThreadFactory().newThread(() -> {
            try {
                Thread.sleep(3000);
            } catch (InterruptedException ignored) {
            } finally {
                HeroAircraftFactory.getInstance().decreaseShootNum();
            }
        }).start();
    }
}
